input/kafkamdm/lag_monitor.go

```diff
@@ -50,71 +65,45 @@ func (l *lagLogger) Min() int {
 	}
 	min := -1
 	for _, m := range l.measurements {
-		if min < 0 || m < min {
-			min = m
+		lag := int(m.highWaterMark - m.readOffset)
+		if min < 0 || lag < min {
+			min = lag
 		}
 	}
 	return min
 }
 
-type rateLogger struct {
-	lastOffset int64
-	lastTs     time.Time
-	rate       int64
-}
+// Rate returns an average rate calculation over the last N measurements (configured at construction)
+// after startup, reported rate may be 0 if we haven't been up long enough to determine it yet.
+func (l *lagLogger) Rate() int64 {
+	if len(l.measurements) < 2 {
+		return 0
+	}
 
-func newRateLogger() *rateLogger {
-	return &rateLogger{}
-}
+	latestLag := l.measurements[l.pos]
+	var earliestLag offsetMeasurement
 
-// Store saves the current offset and updates the rate if it is confident
-// offset must be concrete values, not logical values like -2 (oldest) or -1 (newest)
-func (o *rateLogger) Store(offset int64, ts time.Time) {
-	if o.lastTs.IsZero() {
-		// first measurement
-		o.lastOffset = offset
-		o.lastTs = ts
-		return
+	if len(l.measurements) == cap(l.measurements) {
+		earliestLag = l.measurements[(l.pos+1)%len(l.measurements)]
+	} else {
+		earliestLag = l.measurements[0]
 	}
-	duration := ts.Sub(o.lastTs)
-	if duration < time.Second && duration > 0 {
-		// too small difference. either due to clock adjustment or this method
-		// is called very frequently, e.g. due to a subsecond offset-commit-interval.
-		// We need to let more time pass to make an accurate calculation.
-		return
-	}
-	if duration <= 0 {
-		// current ts is <= last ts. This would only happen if clock went back in time
-		// in which case we can't reliably work out how
-		// long it has really been since we last took a measurement.
-		// but set a new baseline for next time
-		o.lastTs = ts
-		o.lastOffset = offset
-		return
+
+	high, low := latestLag.highWaterMark, earliestLag.highWaterMark
+
+	// If there are no longer any incoming messages, use our read offsets to compute rate
+	if high == low {
+		high, low = latestLag.readOffset, earliestLag.readOffset
```

Comment:
What's the reason why we wouldn't just always use the `.readOffset`, instead of only using it if the `.highWaterMarks` are equal?
If I understand this correctly then it will now calculate the rate based on how fast Kafka receives new messages in this partition, unless it receives no new messages, then it will calculate the rate based on how fast MT consumes them, right? It seems a little surprising to me that the caller of `.Rate()` doesn't know which of those two rates it will actually get returned...

Reply:
The goal is to estimate the interval between now and the time this message was put into kafka. Using the rate of growth of the partition offsets in kafka is better for this than our read offset, since various things can impact our ingest rate (GC, Prune locks, etc). The old implementation relied on the read offset and was susceptible to over correction due to short pauses in ingest.

As for the failover...it could be better, and I'm open to suggestions. I think that @Dieterbe was looking into a more complete refactor that split the computations from the usage.

I'm most interested in coming up with the most effective heuristic for estimating eligibility in cluster processing. If that means that a well-known fallback behavior is used, then so be it.

Reply:
Ok, makes sense :+1: